### PR TITLE
Rust Coding Guidelines

### PR DESCRIPTION
**Pull request by zi0Black**
Added:
- RUST_CODING_STYLE.md
- RUST_SECURE_CODING.md

Updated:
- CONTRIBUTING.md

### Description
Tentative to bring back the Rust coding guidelines for the Aptos Core

### Test Plan
N/A

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -17,7 +17,8 @@ For development environment setup and first build, see [Building Aptos From Sour
 
 ### Code Style, Hints, and Testing
 
-Refer to our Coding Guidelines for the [Move](./developer-docs-site/docs/move/book/coding-conventions.md) and [Rust](./developer-docs-site/docs/contribute/rust-coding-guidelines.md) programming languages for detailed guidance about how to contribute to the project.
+Refer to our Coding Guidelines for the [Move](https://aptos.dev/move/book/coding-conventions/) and [Rust](./RUST_CODING_STYLE.md) programming languages for detailed guidance about how to contribute to the project.
```
Comment (gedigi) [inline CONTRIBUTING.md:20]:
```suggestion
Refer to our Coding Guidelines for the [Move](https://aptos.dev/move/book/coding-conventions/) and [Rust](./RUST_CODING_STYLE.md) programming languages for detailed guidance about how to contribute to the project.

Also, please ensure you follow our [Secure Coding Guidelines](./RUST_SECURE_CODING.md) to contribute to Aptos securely. 
```

File: RUST_CODING_STYLE.md [hunk 2]
```diff
@@ -0,0 +1,364 @@
+# Coding Guidelines for Aptos Core
+
+This document describes the coding guidelines for the [Aptos Core](https://github.com/aptos-labs/aptos-core) Rust codebase. For the Move language, see the [Move Coding Conventions](https://aptos.dev/move/book/SUMMARY).
+Secure coding guidance is provided in the [Aptos Rust Secure Coding Guidelines](./RUST_SECURE_CODING.md).
```
Comment (gedigi) [inline RUST_CODING_STYLE.md:4]: I suggest we move this into `CONTRIBUTING.md` to give it more relevance and visibility.

File: RUST_SECURE_CODING.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+## Types Systems and Data Structures
+
+### Drop Trait
+
+Implement the `Drop` trait selectively, only when necessary for specific destructor logic. It's mainly used for managing external resources or memory in structures like Box or Rc, often involving unsafe code and security-critical operations.
+
+In a Rust secure development, the implementation of the `std::ops::Drop` trait
+must not panic.
+
+> Do not rely on `Drop` trait in security material treatment after the use, use [zeroize](https://docs.rs/zeroize/latest/zeroize/#) to explicit destroy security material, e.g. private keys.
```
Comment (gedigi) [inline RUST_SECURE_CODING.md:96]: why is this quoted?

same for the other two lines in the doc

File: RUST_SECURE_CODING.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+> Reference cycles can also cause memory leakage [[Rustbook: leak]](https://doc.rust-lang.org/book/ch15-06-reference-cycles.html?highlight=leak#reference-cycles-can-leak-memory).
+
+Most memory leaks result in general product reliability problems. If an attacker can intentionally trigger a memory leak, the attacker might be able to launch a denial-of-service attack (by crashing or hanging the program).
+
+### Fuzzing
+
+Aptos contains harnesses for fuzzing crash-prone code like deserializers, using [`libFuzzer`](https://llvm.org/docs/LibFuzzer.html) through [`cargo fuzz`](https://rust-fuzz.github.io/book/cargo-fuzz.html). For more examples, see the `testsuite/fuzzer` directory where find detailed README.md.
+
+## Appendix
```
Comment (gedigi) [inline RUST_SECURE_CODING.md:161]: we can probably remove this as there's no appendix content

File: RUST_SECURE_CODING.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+#### References
```
Comment (gedigi) [inline RUST_SECURE_CODING.md:163]: this should probably be after `Conclusion` as `References` are commonly added after the content. possibly with a lower indent?

Comment (gedigi) [inline RUST_SECURE_CODING.md:163]:
```suggestion
### References
```

File: RUST_SECURE_CODING.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+- ANSSI's Secure Rust Guidelines: https://anssi-fr.github.io/rust-guide/
+
+# Conclusion
```
Comment (gedigi) [inline RUST_SECURE_CODING.md:167]:
```suggestion
## Conclusion
```

File: RUST_CODING_STYLE.md [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+All code formatting is enforced with [rustfmt](https://github.com/rust-lang/rustfmt) with a project-specific configuration. Below is an example command to adhere to the Aptos Core project conventions.
+
+```
+cargo fmt
+```
+
+## Code analysis
+
+[Clippy](https://github.com/rust-lang/rust-clippy) is used to catch common mistakes and is run as a part of continuous integration. Before submitting your code for review, you can run clippy with our configuration:
+
```
Comment (gregnazario) [inline RUST_CODING_STYLE.md:17]: Note that there is a script to handle all of these, so are these rust guidelines for this repository or for anything in Rust?

Comment (zi0Black) [reply RUST_CODING_STYLE.md:17]: I copied this from the actual DOC, probably it wasn't updated. Let me fix this :)

**Review by GotenJBZ — APPROVED** [review]

**Pull request merged**